pkg/storage/raft_log_queue.go

```diff
@@ -388,17 +388,28 @@ func computeTruncateDecision(input truncateDecisionInput) truncateDecision {
 
 	// If new first index dropped below first index, make them equal (resulting
 	// in a no-op).
-	if decision.NewFirstIndex < decision.Input.FirstIndex {
-		decision.NewFirstIndex = decision.Input.FirstIndex
+	if decision.NewFirstIndex < input.FirstIndex {
+		decision.NewFirstIndex = input.FirstIndex
 		decision.ChosenVia = truncatableIndexChosenViaFirstIndex
 	}
+
+	// Invariants: NewFirstIndex >= FirstIndex
+	//             NewFirstIndex <= LastIndex (if != 10)
```

Comment:
> I certainly wouldn't want it to return a truncate decision that has NewFirstIndex > LastIndex

This already does happen (and why my first revision failed teamcity). `input.FirstIndex` is set to [TruncatedState.Index + 1](https://github.com/cockroachdb/cockroach/blob/9d62feaf2122872c5584ed0ff4ba1dafe88e7cb6/pkg/storage/replica_raftstorage.go#L347), so 11 for uinit'ed replicas, whereas LastIndex is 10.

> The code in line 384 above seems like it would set NewFirstIndex := 10 in this case. Is that not what happens?

It [does](https://github.com/cockroachdb/cockroach/blob/a2948d0be761db5940fc67d420eb52f1efef7c99/pkg/storage/raft_log_queue.go#L384-L387), but given 10 < 11 (input.FirstIndex), it's [brought back up](https://github.com/cockroachdb/cockroach/blob/a2948d0be761db5940fc67d420eb52f1efef7c99/pkg/storage/raft_log_queue.go#L389-L394) to 11. And thus we have NewFirstIndex > LastIndex. So this is funky, only for uninit'ed replicas can we have input.FirstIndex > input.LastIndex (tnput.FirstIndex = input.LastIndex + 1).